[FLINK-14646] Check non-null for key in KeyGroupStreamPartitioner

buptljy opened the pull request:
## What is the purpose of the change

Add non-null check for key in KeyGroupStreamPartitioner because the NPE message thrown is not friendly to users.

## Brief change log

Add Preconditions to check null.

buptljy commented:
cc @tillrohrmann

tillrohrmann commented on flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/partitioner/KeyGroupStreamPartitioner.java:
```diff
@@ -55,6 +55,7 @@ public int selectChannel(SerializationDelegate<StreamRecord<T>> record) {
 		} catch (Exception e) {
 			throw new RuntimeException("Could not extract key from " + record.getInstance().getValue(), e);
 		}
+		Preconditions.checkNotNull(key, "HASH key must not be null!");
```
Should we move this check into the `KeyGroupRangeAssignment`? Then every caller can benefit from them.

buptljy replied:
I kind of misunderstood the sentence `Fields, parameters, and return types are always non-null, unless indicated otherwise` when I read Code Style Guide yesterday :). You're right, it'd be better to move the check into `KeyGroupRangeAssignment`.

buptljy commented:
@tillrohrmann Thanks for your review.  I've addressed your comments and also add a `@Nullable` annotation on the function parameter. Take a look when you spare time. :-)

tillrohrmann commented on flink-runtime/src/main/java/org/apache/flink/runtime/state/KeyGroupRangeAssignment.java:
```diff
@@ -45,7 +47,8 @@ private KeyGroupRangeAssignment() {
 	 * @return the index of the parallel operator to which the given key should be routed.
 	 */
-	public static int assignKeyToParallelOperator(Object key, int maxParallelism, int parallelism) {
+	public static int assignKeyToParallelOperator(@Nullable Object key, int maxParallelism, int parallelism) {
```
We should only annotate fields/parameters with `@Nullable` if we allow them to be `null`. This is here not the case.

buptljy commented:
@tillrohrmann Get it! Thanks for your explaination.
